# Coding Exercises (Part 1)

Now, you will have the opportunity to practice what you have learned. <br>
__Follow the instructions__ and insert your code! 

The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need some help or if you want to check your code, you can also have a look at the __solutions__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Exercise 7: The Numpy Package: Working with numbers made easy!

1. Import Numpy!

In [1]:
import numpy as np

2. You have the following stock Prices in EUR Currency : [23.65, 43.21, 73.01, 87.32]<br>

Create a __Numpy array__ and __convert all prices__ to USD Currency @ an fx rate of __1.2__ (1.2 USD per 1 EUR).

In [4]:
price = [23.65, 43.21, 73.01, 87.32]
price_array = np.array(price)
price_array

array([23.65, 43.21, 73.01, 87.32])

In [7]:
price_array * 1.2

array([ 28.38 ,  51.852,  87.612, 104.784])

(result: array([ 28.38 ,  51.852,  87.612, 104.784])

You have the 
- end-of-2017 stock prices for four companies: [33.65, 63.21, 53.01, 17.32]
- end-of-2018 stock prices for the same companies: [36.54, 61.59, 60.31, 10.40]

3. Calculate the __absolute__ price increase / decrease in 2018 __for all four companies__. Use __vectorized__ numpy code!

In [9]:
price1 = np.array([33.65, 63.21, 53.01, 17.32])
price2 = np.array([36.54, 61.59, 60.31, 10.40])

In [12]:
np.abs(price2 / price1 -1)

array([0.0858841 , 0.02562886, 0.13770987, 0.39953811])

4. Calculate the __price return__ in 2018 __for all four companies__. Use __vectorized__ numpy code!

In [13]:
price2 / price1 -1

array([ 0.0858841 , -0.02562886,  0.13770987, -0.39953811])

You have the following list with cashflows (starting today at t0): [-500, -100, 200, 600, 1000, 800]

In [14]:
#run the cell!
l  = [-500, -100, 200, 600, 1000, 800]

5. Convert l into a __numpy array__ and __save__ the array in memory! __Select__ the __second cashflow__ at t1.

In [17]:
l_array = np.array(l)
l_array[1]

-100

6. __Slice__ the array for the __last two cashflows__!

In [21]:
l_array[-2:]

array([1000,  800])

7. __Slice__ the array for the cashflows in __t2, t3 and t4__.

In [51]:
l_array[2:5]

array([ 200,  600, 1000])

The ABC Company evaluates to buy an additional machine that will increase future profits/cashflows by
- 50 USD in t1, 
- 80 USD in t2,
- 120 USD in t3, 
- 150 USD in t4,
- 200 USD in t5. (each cf at period´s end)<br>

8. The machine costs 500 USD (Investment in to). Calculate the __Project´s NPV__ and evaluate whether ABC should pursue the project. <br> ABC´s required rate of return (Cost of Capital) is 7.5% p.a. Use __vectorized__ numpy code!

In [24]:
# run the cell
cf = np.array([-500, 50, 80, 120, 150, 200])
cf

array([-500,   50,   80,  120,  150,  200])

In [33]:
r = 0.075
f = 1+r
#define n periods as an array so we can operate on it
n = np.array([0, 1, 2, 3, 4, 5])

In [34]:
pv_array = cf / f**n

npv = pv_array.sum()
npv

-36.03468883756554

(result: -36.03468883756554 -> no!)

9. __Slice__ the array cf (from question 8) for the positive inflows (__from t1__) and __save__ the slice in the __variable cf_pos__. <br> Make sure that once you __change the slice cf_pos__, the array __cf shall change as well__!

In [52]:
cf_pos = cf[1:]
cf_pos

array([ 50,  80, 120, 150, 200])

10. Change the first cashflow (t1) of __cf_pos to 70__. Check whether __cf gets mutated as well__!

In [54]:
cf_pos[0] = 70
cf

array([-500,   70,   80,  120,  150,  200])

11. __Restore cf__ and __redo Q9 and Q10__. This time make sure that cf and cf_pos are __independent objects__! 

In [55]:
cf = np.array([-500, 50, 80, 120, 150, 200])
cf

array([-500,   50,   80,  120,  150,  200])

In [56]:
#use .copy() to create indepdenent objects
cf_pos = cf[1:].copy()
cf_pos

array([ 50,  80, 120, 150, 200])

In [57]:
cf_pos[0] = 70
cf_pos

array([ 70,  80, 120, 150, 200])

12. Calculate the __total sum__ over all cashflows of cf!

In [35]:
# run the cell
cf = np.array([-500, 50, 80, 120, 150, 200])
cf

array([-500,   50,   80,  120,  150,  200])

In [38]:
cf.sum()

100

13. Return the __highest Value__ of cf!

In [39]:
cf.max()

200

14. Calculate the __cumulative cashflows__ for cf!

In [40]:
cf.cumsum()

array([-500, -450, -370, -250, -100,  100], dtype=int32)

15. Assume the following list: [3, 10, 6, 45, 100]. Calculate the __square root__ for all elements in the list. 

In [45]:
num = np.array([3, 10, 6, 45, 100])
np.sqrt(num)

array([ 1.73205081,  3.16227766,  2.44948974,  6.70820393, 10.        ])

Assume the following stock prices:

In [46]:
# run the cell!
a = np.array([0.34, 43.56, 32.12, 67.43, 0.87, 9.87])
a

array([ 0.34, 43.56, 32.12, 67.43,  0.87,  9.87])

16. __Filter__ a for all penny stocks! (price less than 1)

In [48]:
mask1 = a[a < 1] 
mask1

array([0.34, 0.87])

(result: array([0.34, 0.87]))

17. __Filter__ a for all stocks prices that are greater than 10 __and__ less than 50 (logical and)!

In [58]:
a[(a > 10) & (a < 50)]

array([43.56, 32.12])

18. __Filter__ a for stocks that are either penny stocks __or__ have prices greater than 60 (logical or)!

In [59]:
a[(a < 1) | (a > 60)]

array([ 0.34, 67.43,  0.87])

(result: array([ 0.34, 67.43,  0.87]))

19. Get the __reverse__ of Q18! (Neither penny stocks nor prices greater than 60)

In [60]:
a[~((a < 1) | (a  > 60))]

array([43.56, 32.12,  9.87])

(result: array([43.56, 32.12,  9.87]))

20. Create a ___numpy array__ with interest rates from __0% till 10%__ (incl.) and a __step of 0.5%__. 

In [63]:
#starting point, top of range (exclusive), step
np.arange(0, 0.101, 0.005)

array([0.   , 0.005, 0.01 , 0.015, 0.02 , 0.025, 0.03 , 0.035, 0.04 ,
       0.045, 0.05 , 0.055, 0.06 , 0.065, 0.07 , 0.075, 0.08 , 0.085,
       0.09 , 0.095, 0.1  ])

21. __Plot__ the f(x) = cos(x) in the interval [-10, 10]! (use 1,000 data points)

In [64]:
#create 1000 evenly spaced numbers for x axis with np.linspace()
x = np.linspace(-10, 10, 1000)
y = np.cos(x)

In [68]:
#import matplotlib
import matplotlib as plt

In [71]:
plt.figure(figsize = (12, 8))
plt.plot(x, y)
plt.title("f(x) = cos(x)", fontsize = 20)
plt.xlabel("x", fontsize = 15)
plt.ylabel("y", fontsize = 15)
plt.show()

AttributeError: module 'matplotlib' has no attribute 'figure'

# Well Done!

# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

1. Import Numpy!

In [ ]:
import numpy as np

2. You have the following stock Prices in EUR Currency : [23.65, 43.21, 73.01, 87.32]<br>

Create a __Numpy array__ and __convert all prices__ to USD Currency @ an fx rate of __1.2__ (1.2 USD per 1 EUR).

In [ ]:
a = np.array([23.65, 43.21, 73.01, 87.32])
a

In [ ]:
a * 1.2

(result: array([ 28.38 ,  51.852,  87.612, 104.784])

You have the 
- end-of-2017 stock prices for four companies: [33.65, 63.21, 53.01, 17.32]
- end-of-2018 stock prices for the same companies: [36.54, 61.59, 60.31, 10.40]

3. Calculate the __absolute__ price increase / decrease in 2018 __for all four companies__. Use __vectorized__ numpy code!

In [ ]:
p_2017 = np.array([33.65, 63.21, 53.01, 17.32])
p_2018 = np.array([36.54, 61.59, 60.31, 10.40])

In [ ]:
p_2018 - p_2017

4. Calculate the __price return__ in 2018 __for all four companies__. Use __vectorized__ numpy code!

In [ ]:
(p_2018 / p_2017) -1 

You have the following list with cashflows (starting today at t0): [-500, -100, 200, 600, 1000, 800]

In [ ]:
#run the cell!
l  = [-500, -100, 200, 600, 1000, 800]

5. Convert l into a __numpy array__ and __save__ the array in memory! __Select__ the __second cashflow__ at t1.

In [ ]:
a = np.array(l)
a

In [ ]:
a[1]

6. __Slice__ the array for the __last two cashflows__!

In [ ]:
a[-2:]

7. __Slice__ the array for the cashflows in __t2, t3 and t4__.

In [ ]:
a[2:5]

The ABC Company evaluates to buy an additional machine that will increase future profits/cashflows by
- 50 USD in t1, 
- 80 USD in t2,
- 120 USD in t3, 
- 150 USD in t4,
- 200 USD in t5. (each cf at period´s end)<br>

In [ ]:
# run the cell
cf = np.array([-500, 50, 80, 120, 150, 200])
cf

8. The machine costs 500 USD (Investment in to). Calculate the __Project´s NPV__ and evaluate whether ABC should pursue the project. <br> ABC´s required rate of return (Cost of Capital) is 7.5% p.a. Use __vectorized__ numpy code!

In [ ]:
r = 0.075
f = 1 + r

In [ ]:
n = np.array([0, 1, 2, 3, 4, 5])
n

In [ ]:
(cf / f**n).sum()

(result: -36.03468883756554 -> no!)

9. __Slice__ the array cf (from question 8) for the positive inflows (__from t1__) and __save__ the slice in the __variable cf_pos__. <br> Make sure that once you __change the slice cf_pos__, the array __cf shall change as well__!

In [ ]:
cf_pos = cf[1:]
cf_pos

10. Change the first cashflow (t1) of __cf_pos to 70__. Check whether __cf gets mutated as well__!

In [ ]:
cf_pos[0] = 70

In [ ]:
cf

11. __Restore cf__ and __redo Q9 and Q10__. This time make sure that cf and cf_pos are __independent objects__! 

In [ ]:
cf = np.array([-500, 50, 80, 120, 150, 200])
cf

In [ ]:
cf_pos = cf[1:].copy()
cf_pos

In [ ]:
cf_pos[0] = 70
cf_pos

In [ ]:
cf

12. Calculate the __total sum__ over all cashflows of cf!

In [ ]:
# run the cell
cf = np.array([-500, 50, 80, 120, 150, 200])
cf

In [ ]:
cf.sum()

13. Return the __highest Value__ of cf!

In [ ]:
cf.max()

14. Calculate the __cumulative cashflows__ for cf!

In [ ]:
cf.cumsum()

15. Assume the following list: [3, 10, 6, 45, 100]. Calculate the __square root__ for all elements in the list. 

In [ ]:
np.sqrt([3, 10, 6, 45, 100])

Assume the following stock prices:

In [ ]:
# run the cell!
a = np.array([0.34, 43.56, 32.12, 67.43, 0.87, 9.87])
a

16. __Filter__ a for all penny stocks! (price less than 1)

In [ ]:
a[a < 1]

(result: array([0.34, 0.87]))

17. __Filter__ a for all stocks prices that are greater than 10 __and__ less than 50 (logical and)!

In [ ]:
a[(a > 10) & (a < 50)]

18. __Filter__ a for stocks that are either penny stocks __or__ have prices greater than 60 (logical or)!

In [ ]:
a[(a < 1) | (a > 60)]

(result: array([ 0.34, 67.43,  0.87]))

19. Get the __reverse__ of Q18! (Neither penny stocks nor prices greater than 60)

In [ ]:
a[~((a < 1) | (a > 60))]

(result: array([43.56, 32.12,  9.87]))

20. Create a ___numpy array__ with interest rates from __0% till 10%__ (incl.) and a __step of 0.5%__. 

In [ ]:
np.arange(0, 0.101, 0.005)

21. __Plot__ the f(x) = cos(x) in the interval [-10, 10]! (use 1,000 data points)

In [ ]:
x = np.linspace(-10 , 10, 1000)

In [ ]:
y = np.cos(x)

In [ ]:
import matplotlib.pyplot as plt

In [ ]:
plt.figure(figsize = (12, 8))
plt.plot(x, y)
plt.title("f(x) = cos(x)", fontsize = 20)
plt.xlabel("x", fontsize = 15)
plt.ylabel("y", fontsize = 15)
plt.show()